src/Http/Shared/PipeCopyOperationInternal.cs

```diff
@@ -0,0 +1,87 @@
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http
+{
+    // FYI: In most cases the source will be a FileStream and the destination will be to the network.
+    internal static class PipeCopyOperationInternal
+    {
+        private const int DefaultBufferSize = 4096;
+
+        /// <summary>Asynchronously reads the given number of bytes from the source stream and writes them using pipe writer.</summary>
+        /// <returns>A task that represents the asynchronous copy operation.</returns>
+        /// <param name="source">The stream from which the contents will be copied.</param>
+        /// <param name="writer"></param>
+        /// <param name="count">The count of bytes to be copied.</param>
+        /// <param name="cancel">The token to monitor for cancellation requests.</param>
+        public static Task CopyToAsync(Stream source, PipeWriter writer, long? count, CancellationToken cancel)
+        {
+            return CopyToAsync(source, writer, count, DefaultBufferSize, cancel);
+        }
+
+        /// <summary>Asynchronously reads the given number of bytes from the source stream and writes them using pipe writer.</summary>
+        /// <returns>A task that represents the asynchronous copy operation.</returns>
+        /// <param name="source">The stream from which the contents will be copied.</param>
+        /// <param name="writer"></param>
+        /// <param name="count">The count of bytes to be copied.</param>
+        /// <param name="bufferSize">The size, in bytes, of the buffer. This value must be greater than zero. The default size is 4096.</param>
+        /// <param name="cancel">The token to monitor for cancellation requests.</param>
+        public static async Task CopyToAsync(Stream source, PipeWriter writer, long? count, int bufferSize, CancellationToken cancel)
+        {
+            long? bytesRemaining = count;
+
+            Debug.Assert(source != null);
+            Debug.Assert(writer != null);
+            Debug.Assert(!bytesRemaining.HasValue || bytesRemaining.GetValueOrDefault() >= 0);
+
+            while (true)
+            {
+                // The natural end of the range.
+                if (bytesRemaining.HasValue && bytesRemaining.GetValueOrDefault() <= 0)
+                {
+                    return;
+                }
+
+                cancel.ThrowIfCancellationRequested();
+
+                var readLength = bufferSize;
+                if (bytesRemaining.HasValue)
+                {
+                    readLength = (int)Math.Min(bytesRemaining.GetValueOrDefault(), readLength);
+                }
+
+                var memory = writer.GetMemory(readLength);
+                var read = await source.ReadAsync(memory, cancel);
+
+                if (bytesRemaining.HasValue)
```

Comment:
I think I fixed this with `memory.Slice(0, readLength)`, but I'm not sure if that's the best solution. Does `Slice` has performance cost?

Reply:
Slicing Memory can have some overhead so you'd want to avoid doing it on every call. Rather than slicing I think you can do a Min check on `read` and `bytesRemaining` and pass the result to Advance.

Reply:
Slicing doesn't have a ton of overhead, but it would still be better not to slice when `memory.Length` is less than `bytesRemaining` or when `bytesRemaining` is null.

I think slicing only when `memory.Length` is greater than `bytesRemaining` and continuing to advance by `read` is fine. It allows `source.ReadAsync` to copy less on the final read.